File: app/src/main/java/social/entourage/android/api/model/EntourageUser.java
package social.entourage.android.api.model;

import androidx.annotation.NonNull;

import com.google.gson.annotations.Expose;
import com.google.gson.annotations.SerializedName;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;

import social.entourage.android.api.model.feed.FeedItem;

/**
 * Created by mihaiionescu on 24/02/16.
 */
public class EntourageUser extends TimestampedObject implements Serializable {

    private static final long serialVersionUID = 6896801856163434601L;

    private final static String HASH_STRING_HEAD = "TourUser-";

    @SerializedName("id")
    private int userId;

    @SerializedName("display_name")
    private String displayName;

    @Expose(serialize = false, deserialize = true)
    private String email;

    @Expose(serialize = false, deserialize = true)
    private String status;

    @SerializedName("requested_at")
    private Date requestDate;

    private String message;

    @SerializedName("avatar_url")
    private String avatarURLAsString;

    @SerializedName("partner")
    private Partner partner;

    @SerializedName("group_role")
    private String groupRole;

    @SerializedName("community_roles")
    private List<String> communityRoles;

    private boolean isDisplayedAsMember = false;

    @Expose(serialize = false, deserialize = false)
    private FeedItem feedItem;

    public String getEmail() {
        return email;
    }

    public void setEmail(final String email) {
        this.email = email;
    }

    public String getDisplayName() {
        return displayName == null ? "" : displayName;
    }

    public void setDisplayName(final String displayName) {
        this.displayName = displayName;
    }

    public Date getRequestDate() {
        return requestDate;
    }

    public void setRequestDate(final Date requestDate) {
        this.requestDate = requestDate;
    }

    public String getStatus() {
        return status;
    }

    public void setStatus(final String status) {
        this.status = status;
    }

    public int getUserId() {
        return userId;
    }

    public void setUserId(final int userId) {
        this.userId = userId;
    }

    public String getMessage() {
        return message;
    }

    public void setMessage(final String message) {
        this.message = message;
    }

    public String getAvatarURLAsString() {
        return avatarURLAsString;
    }

    public void setAvatarURLAsString(final String avatarURLAsString) {
        this.avatarURLAsString = avatarURLAsString;
    }

    public Partner getPartner() {
        return partner;
    }

    public String getGroupRole() {
        return groupRole;
    }

    public List<String> getCommunityRoles() {
        if (communityRoles != null) {
            return communityRoles;
        } else {
            return new ArrayList<>();
        }
    }

    public boolean isDisplayedAsMember() {
        return isDisplayedAsMember;
    }

    public void setDisplayedAsMember(final boolean displayedAsMember) {
        isDisplayedAsMember = displayedAsMember;
    }

    public FeedItem getFeedItem() {
        return feedItem;
    }

    public void setFeedItem(final FeedItem feedItem) {
        this.feedItem = feedItem;
    }

    @Override
    public Date getTimestamp() {
        return requestDate;
    }

    @NonNull
    @Override
    public String hashString() {
        return HASH_STRING_HEAD + userId;
    }

    @Override
    public int getType() {
        return isDisplayedAsMember ? FEED_MEMBER_CARD : TOUR_USER_JOIN;
    }

    @Override
    public long getId() {
        return userId;
    }

    @Override
    public boolean equals(final Object o) {
        if (o == null || o.getClass() != this.getClass()) return false;
        //return (this.userId == ((TourUser)o).userId) && (this.status.equals(((TourUser)o).status));
        return this.userId == ((EntourageUser)o).userId;
    }

    @NonNull
    public EntourageUser clone() {
        EntourageUser clone = new EntourageUser();
        clone.userId = this.userId;
        clone.feedItem = this.feedItem;
        clone.displayName = this.displayName;
        clone.email = this.email;
        clone.status = this.status;
        clone.requestDate = this.requestDate;
        clone.message = this.message;
        clone.avatarURLAsString = this.avatarURLAsString;
        clone.partner = this.partner;
        clone.groupRole = this.groupRole;
        clone.communityRoles = this.communityRoles;
        clone.isDisplayedAsMember = this.isDisplayedAsMember;

        return clone;
    }

    // ----------------------------------
    // INNER CLASSES
    // ----------------------------------

    public static class EntourageUserWrapper {

        private EntourageUser user;

        public EntourageUser getUser() {
            return user;
        }

        public void setUser(EntourageUser user) {
            this.user = user;
        }

    }

    public static class EntourageUserResponse {
        private EntourageUser user;
        public EntourageUser getUser() {
            return user;
        }
        public void setUser(EntourageUser user) {
            this.user = user;
        }
    }

    public static class EntourageUserListResponse {

        private List<EntourageUser> users;

        public List<EntourageUser> getUsers() {
            return users;
        }

        public void setUsers(List<EntourageUser> users) {
            this.users = users;
        }

    }
}

File: app/src/main/java/social/entourage/android/api/model/EntourageUser.kt
package social.entourage.android.api.model

import com.google.gson.annotations.Expose
import com.google.gson.annotations.SerializedName
import social.entourage.android.api.model.feed.FeedItem
import java.io.Serializable
import java.util.Date

/**
 * Created by mihaiionescu on 24/02/16.
 */
class EntourageUser : TimestampedObject(), Serializable {
    @SerializedName("id")
    var userId = 0

    @SerializedName("display_name")
    var displayName: String? = null

    @SerializedName("requested_at")
    private var requestDate: Date? = null
    var message: String? = null

    @SerializedName("confirmed_at")
    var confirmedAt : Boolean? = null

    @SerializedName("participate_at")
    var participateAt : Boolean? = null

    @SerializedName("avatar_url")
    var avatarURLAsString: String? = null

    @SerializedName("partner")
    var partner: Partner? = null
        private set

    @SerializedName("group_role")
    var groupRole: String? = null
        private set

    @SerializedName("photo_acceptance")
    var photoAcceptance: Boolean? = null
        private set

    @SerializedName("community_roles")
    var communityRoles: List<String>? = null
        private set
    @SerializedName("roles")
    var roles: List<String>? = null
        private set



    var isDisplayedAsMember = false

    @Expose(serialize = false, deserialize = true)
    private var email: String? = null

    @Expose(serialize = false, deserialize = true)
    var status: String? = null

    @Expose(serialize = false, deserialize = false)
    var feedItem: FeedItem? = null

    var partner_role_title:String? = null

    override val timestamp: Date?
    get() = requestDate

    override fun hashString(): String {
        return HASH_STRING_HEAD + userId
    }

    override val type: Int
        get() = if (isDisplayedAsMember) FEED_MEMBER_CARD else USER_JOIN

    override val id: Long
        get() = userId.toLong()

    override fun equals(other: Any?): Boolean {
        return if (other == null || other.javaClass != this.javaClass) false else userId == (other as EntourageUser).userId
    }

    fun clone(): EntourageUser {
        val clone = EntourageUser()
        clone.userId = userId
        clone.feedItem = feedItem
        clone.displayName = displayName
        clone.email = email
        clone.confirmedAt = confirmedAt
        clone.status = status
        clone.requestDate = requestDate
        clone.message = message
        clone.avatarURLAsString = avatarURLAsString
        clone.partner = partner
        clone.groupRole = groupRole
        clone.communityRoles = communityRoles
        clone.isDisplayedAsMember = isDisplayedAsMember
        return clone
    }

    fun getCommunityRoleWithPartnerFormated() : String? {
        communityRoles?.let { roles->
            var roleStr = ""

            if (isAdmin()) {
                roleStr = "Admin"
            }

            for (role in roles) {
                if(roleStr.isNotEmpty()) {
                    roleStr = "$roleStr • $role"
                }
                else {
                    roleStr = role
                }

                break //TODO: why do we get only the first role ?
            }
            partner?.name?.let { partnerName->
                if (roleStr.isNotEmpty()) {
                    roleStr = "$roleStr • $partnerName"
                }
                else {
                    roleStr = partnerName
                }
            }
            return roleStr
        }


        return null
    }

    fun isAdmin() : Boolean {
        if (groupRole == "creator") {
            return true
        }
        return false
    }

    fun isAmbassador() : Boolean {
        if (communityRoles?.contains("Ambassadeur") == true ||communityRoles?.contains("ambassador") == true) {
            return true
        }
        return false
    }

    companion object {
        private const val serialVersionUID = 6896833312363434601L
        private const val HASH_STRING_HEAD = "EntourageUser-"
    }
}